### PR TITLE
API: Add ``device`` and ``to_device`` to scalars

### PR DESCRIPTION
**Pull request by mtsokol**
Addresses https://github.com/numpy/numpy/issues/26850

Hi @seberg,

This PR adds `device` attribute and `to_device` method to NumPy scalars, as decided in https://github.com/numpy/numpy/issues/26850#issuecomment-2210983792.

As `device`, `to_device`, and `__array_namespace__` members are present in arrays and scalars now, therefore I decided to move them out to `array_api_standard.c` and `.h` files to avoid duplication, is it OK?

For NDArray objects these functions are prefixed with `array_` and for scalars it's `gentype_`. I wonder which prefix to use here to unify these two conventions for these functions.

### CODE REVIEW

File: numpy/__init__.pyi [hunk 1]
```diff
@@ -2578,12 +2581,18 @@ class ndarray(_ArrayOrScalarCommon, Generic[_ShapeType, _DType_co]):
...
+    @overload
+    def to_device(self: NDArray[np.bool], device: L["cpu"], /, *, stream: None | int | Any = ...) -> NDArray[np.bool]: ...
+    @overload
+    def to_device(self: NDArray[_NumberType], device: L["cpu"], /, *, stream: None | int | Any = ...) -> NDArray[_NumberType]: ...
+    @overload
+    def to_device(self: NDArray[timedelta64], device: L["cpu"], /, *, stream: None | int | Any = ...) -> NDArray[timedelta64]: ...
+    @overload
+    def to_device(self: NDArray[object_], device: L["cpu"], /, *, stream: None | int | Any = ...) -> NDArray[object_]: ...
+    @overload
+    def to_device(self: NDArray[Any], device: L["cpu"], /, *, stream: None | int | Any = ...) -> NDArray[Any]: ...
```
Comment (mattip) [inline numpy/__init__.pyi:2595]: Do we really need all these overloads, isn't the last one enough?

File: numpy/_core/src/multiarray/array_api_standard.c [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    if (stream != Py_None) {
+        PyErr_SetString(PyExc_ValueError,
+                        "The stream argument in to_device() "
+                        "is not supported");
+        return NULL;
+    }
+
+    if (strcmp(device, "cpu") != 0) {
+        PyErr_Format(PyExc_ValueError,
+                     "Unsupported device: %s.", device);
```
Comment (mattip) [inline numpy/_core/src/multiarray/array_api_standard.c:35]: Maybe this instead?
```suggestion
                     "device kwarg must be 'cpu' not: %s.", device);
```

File: numpy/__init__.pyi (hunk 1 above)
Comment (mtsokol) [reply numpy/__init__.pyi:2595]: With these overloads we can have more specialized typing, like:
```python
assert_type(AR_f8.to_device("cpu"), npt.NDArray[np.float64])
assert_type(AR_i8.to_device("cpu"), npt.NDArray[np.int64])
```
Similar overloads are present for `__abs__` or `__invert__` methods.

Without these overloads the typing result of the `to_device()` call will always be `NDArray[Any]`.

File: numpy/_core/src/multiarray/array_api_standard.c (hunk 2 above)
Comment (mtsokol) [reply numpy/_core/src/multiarray/array_api_standard.c:35]: I would still use "Unsupported" word here to indicate why the `"cpu"` is only accepted.

How about: `"Unsupported device: %s. Only 'cpu' is accepted."`?

File: numpy/__init__.pyi (hunk 1 above)
Comment (mattip) [reply numpy/__init__.pyi:2595]: OK, but it all seems very verbose to me. Isn't there a more concise way to say that the output dtype is the same as the input dtype?

Comment (mtsokol) [reply numpy/__init__.pyi:2588]: Ah, right. I checked it and it's possible - I reduced it to two overloads: one that says input dtype is the same as output, and one that returns `Any`. I also expanded typing tests for `to_device`.

File: numpy/_core/src/multiarray/array_api_standard.c (hunk 2 above)
Comment (mtsokol) [reply numpy/_core/src/multiarray/array_api_standard.c]: Done!

File: numpy/__init__.pyi (hunk 1 above)
Comment (mattip) [reply numpy/__init__.pyi:2588]: Thanks. Curious: what does `_SCT` stand for?

**Comment by mattip** [comment]
Thanks @mtsokol

**Pull request merged**